A71-xds-fallback.md

```diff
@@ -0,0 +1,135 @@
+A66: xDS Fallback
+----
+* Author(s): @eostroukhov
+* Approver: @markdroth
+* Status: {Draft, In Review, Ready for Implementation, Implemented}
+* Implemented in: <language, ...>
+* Last updated: 2023-07-27
+* Discussion at: <google group thread> (filled after thread exists)
+
+## Abstract
+
+This proposal describes a fallback mechanism for situtaions when the primary
+xDS server is unavailable. xDS support for client-side gRPC is described
+in gRFC [A27][A27] and server-side support is described in [A36][A36].
+
+Several specific scenarios need to be considered:
+1. The xDS server is not available at the time of the initial connection.
+1. The xDS server is no longer available after some resources have been 
+obtained.
+
+## Background
+
+xDS (also known as xDS Discovery Service) is a suite of APIs for discovering
+and subscribing to the configuration of a server mesh. Orignially created by
+for the [Envoy Proxy](http://envoyproxy.io) project, the xDS protocol is
+evolving into an industry standard. gRPC introduced a number of features that
+can be configured using this protocol.
+
+Components of the server mesh communicate with xDS server(s) (also known as
+a "control plane") to obtain the initial configuration and to receive updates
+to the configuration.
+
+An xDS configuration consists of a number of "resources" that can be queried
+separately and may change independently from one another.
+
+Even a brief downtime of the xDS control plane may cause significant disruption
+in the service mesh inter-component connectivity and result in wider outages.
+Examples include:
+
+* Services becoming unavailable as they fail to access downstream dependencies.
+* Traffic being routed incorrectly failing to properly balance the load between
+the endpoints.
+* Security risks may be introduced as outdated credentials are used for
+communications between the servers.
+
+Current xDS implementation in gRPC uses a single global instance of
+the XdsClient class to communicate with the xDS control plane. This instance
+is shared across all channels to reduce overhead by providing a shared cache
+for xDS resources as well as reducing a number of connections to xDS servers.
+
+### Related Proposals: 
+* [A27: xDS-Based Global Load Balancing](A27)
+* [A36: xDS-Enabled Servers](A36)
+* [A57: XdsClient Failure Mode Behavior](A57)
+
+[A27]: A27-xds-global-load-balancing.md
+[A36]: A36-xds-for-servers.md
+[A57]: A57-xds-client-failure-mode-behavior.md
+
+## Proposal
+
+gRPC code should use fallback iff both:
+
+* primary server not reachable
+* at least one watcher exists for a resource that is not cached (where
+   "does not exist" counts as cached).
+
+Instead of using a global XdsClient instance, gRPC will use a shared XdsClient
+instance for each data plane target.  In other words, if two channels are
```

Comment:
What is the benefit of using multiple separate XdsClient instances instead of one global instance?

Reply:
I think this section needs to be fleshed out more.  In particular, here are some key points that I think this needs to cover:
- We are expecting to use this in an environment where the data provided by the secondary xDS server is better than nothing but is less desirable than the data in the primary server.  This means that while we definitely want to use the fallback data in preference to the client simply failing, we will always prefer to use the data from the primary server over the data from the fallback server.  In principle, this means that whenever we have a resource already cached from the primary server, we want to use that instead of falling back.
- We have no guarantee that a combination of resources from the primary and secondary server forms a valid cohesive configuration, so we cannot make this determination on a per-resource basis.  We need any given gRPC channel to either use all resources from the primary server or all resources from the secondary server.
- The ideal behavior is for each individual gRPC channel to fall back independently of any other, but the global XdsClient instance makes that challenging, because we need to make the fallback decision independently for each authority, and that is handles inside the XdsClient.  (And we do still want to share XdsClient instances across channels, both to reduce load on the control plane from duplicate channels and to make CSDS work.)

If we continue to use a singleton shared XdsClient instance, then we will cause individual channels that already have all of their resources cached from the primary server to switch to fallback resources in the following cases:
- If the primary server goes down and a gRPC client then creates a new xDS-enabled channel for a target for which the resources are not already cached, that will trigger the client to fall back, not just for the new channel but for all channels.
- Cases where the primary server goes down while the client is in the middle of processing an update (referred to as the "mid-update" case).  For example, let's say that the XdsClient gets an updated RouteConfig that includes a new cluster that it was not previously using, thus triggering the client to subscribe to the new Cluster resource, but then TD goes down before sending the new Cluster resource.  At this point, the XdsClient will fall back to the secondary server, and it will use the fallback resources instead of the cached resources that it already had from the primary server.

The best compromise that we can find is to use a different XdsClient instance for each target, but still share those XdsClient instances between channels for the same target.  This is not exactly the same as having each individual channel fall back independently, but it's close enough in practice: if there are multiple channels for the same target, they will all use the same decision about whether or not to fall back, but that should be fine, since they are also using the same set of xDS resources.

Reply:
Copied verbatim.